File: android/src/main/java/com/singular/flutter_sdk/SingularSDK.java
package com.singular.flutter_sdk;

import android.content.Context;
import android.content.Intent;
import android.os.Handler;
import android.os.Looper;

import androidx.annotation.NonNull;

import com.singular.sdk.ShortLinkHandler;
import com.singular.sdk.Singular;
import com.singular.sdk.SingularConfig;
import com.singular.sdk.SingularDeviceAttributionHandler;
import com.singular.sdk.SingularLinkHandler;
import com.singular.sdk.SingularLinkParams;

import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

/** FlutterSdkPlugin */
public class SingularSDK implements FlutterPlugin, ActivityAware, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private MethodChannel channel;
  private static Context mContext;
  private Intent mIntent;
  private Handler uiThreadHandler = new Handler(Looper.getMainLooper());
  private static int currentIntentHash = 0;
  private static Map configDict;
  private static SingularLinkHandler singularLinkHandler;
  private static SingularConfig singularConfig;

  public static void onNewIntent(Intent intent) {

    // We save the intent hash code to make sure that the intent we get is a new one to avoid resolving an old deeplink.
    if (singularConfig != null &&
            singularLinkHandler != null && intent != null && intent.hashCode() != currentIntentHash && intent.getData() != null &&
            Intent.ACTION_VIEW.equals(intent.getAction())) {
      currentIntentHash = intent.hashCode();

      singularConfig.withSingularLink(intent, singularLinkHandler);
      Singular.init(mContext, singularConfig);
    }
  }

  // Notify the plugin that it has been attached to an engine.
  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "singular-api");
    channel.setMethodCallHandler(this);

    mContext = flutterPluginBinding.getApplicationContext();
  }

  // ActivityAware
  @Override
  public void onAttachedToActivity(ActivityPluginBinding binding) {
    mIntent = binding.getActivity().getIntent();
  }

  @Override
  public void onDetachedFromActivityForConfigChanges() {
  }

  @Override
  public void onReattachedToActivityForConfigChanges(
          ActivityPluginBinding binding) {
  }

  @Override
  public void onDetachedFromActivity() {
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {

    switch (call.method) {
      case SingularConstants.START:
        start(call, result);
        break;
      case SingularConstants.SET_CUSTOM_USER_ID:
        setCustomUserId(call, result);
        break;
      case SingularConstants.UNSET_CUSTOM_USER_ID:
        unsetCustomUserId(call, result);
        break;
      case SingularConstants.SET_DEVICE_CUSTOM_USER_ID:
        setDeviceCustomUserId(call, result);
        break;
      case SingularConstants.EVENT:
        event(call, result);
        break;
      case SingularConstants.EVENT_WITH_ARGS:
        eventWithArgs(call, result);
        break;
      case SingularConstants.CUSTOM_REVENUE:
        customRevenue(call, result);
        break;
      case SingularConstants.CUSTOM_REVENUE_WITH_ATTRIBUTES:
        customRevenueWithArgs(call, result);
        break;
      case SingularConstants.TRACKING_OPT_IN:
        trackingOptIn(call, result);
        break;
      case SingularConstants.TRACKING_UNDER13:
        trackingUnder13(call, result);
        break;
      case SingularConstants.STOP_ALL_TRACKING:
        stopAllTracking(call, result);
        break;
      case SingularConstants.RESUME_ALL_TRACKING:
        resumeAllTracking(call, result);
        break;
      case SingularConstants.IS_ALL_TRACKING_STOPPED:
        isAllTrackingStopped(call, result);
        break;
      case SingularConstants.LIMIT_DATA_SHARING:
        limitDataSharing(call, result);
        break;
      case SingularConstants.GET_LIMIT_DATA_SHARING:
        getLimitDataSharing(call, result);
        break;
      case SingularConstants.CLEAR_GLOBAL_PROPERTIES:
        clearGlobalProperties(call, result);
        break;
      case SingularConstants.UNSET_GLOBAL_PROPERTY:
        unsetGlobalProperty(call, result);
        break;
      case SingularConstants.SET_GLOBAL_PROPERTY:
        setGlobalProperty(call, result);
        break;
      case SingularConstants.GET_GLOBAL_PROPERTIES:
        getGlobalProperties(call, result);
        break;
      case SingularConstants.SET_WRAPPER_NAME_AND_VERSION:
        setWrapperNameAndVersion(call, result);
        break;
      case SingularConstants.SET_FCM_TOKEN:
        setFCMDeviceToken(call, result);
        break;
      case SingularConstants.REGISTER_DEVICE_TOKEN_FOR_UNINSTALL:
        registerDeviceTokenForUninstall(call, result);
        break;
      case SingularConstants.CREATE_REFERRER_SHORT_LINK:
        createReferrerShortLink(call, result);
        break;
      default:
        result.notImplemented();
        break;
    }
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    if (channel != null) {
      channel.setMethodCallHandler(null);
      channel = null;
    }
    mContext = null;
  }

  void initSDK() {
    if (configDict == null) {
      return;
    }

    String apiKey = (String) configDict.get("apiKey");
    String secretKey = (String) configDict.get("secretKey");
    boolean collectOAID = (boolean) configDict.get("collectOAID");
    boolean enableLogging = (boolean) configDict.get("enableLogging");

    double shortLinkResolveTimeOut = (double) configDict.get("shortLinkResolveTimeOut");

    singularConfig = new SingularConfig(apiKey, secretKey);

    String customUserId = (String) configDict.get("customUserId");

    if (customUserId != null) {
      singularConfig.withCustomUserId(customUserId);
    }
    if (collectOAID) {
      singularConfig.withOAIDCollection();
    }
    if (enableLogging) {
      singularConfig.withLoggingEnabled();
    }

    try {
      int logLevel = (int) configDict.get("logLevel");

      if (logLevel >= 0) {
        singularConfig.withLogLevel(logLevel);
      }
    } catch (Throwable t) { }

    double sessionTimeout = (double) configDict.get("sessionTimeout");
    if (sessionTimeout >= 0) {
      singularConfig.withSessionTimeoutInSec((long) sessionTimeout);
    }

    Object limitDataSharing = configDict.get("limitDataSharing");
    if (limitDataSharing != null) {
      singularConfig.withLimitDataSharing((boolean) limitDataSharing);
    }

    String imei = (String) configDict.get("imei");
    if (imei != null) {
      singularConfig.withIMEI(imei);
    }

    String facebookAppId = (String) configDict.get("facebookAppId");
    if (facebookAppId != null) {
      singularConfig.withFacebookAppId(facebookAppId);
    }

    try {
      List<String> espDomains = (ArrayList<String>) configDict.get("espDomains");
      if (espDomains != null && espDomains.size() > 0) {
        singularConfig.withESPDomains(espDomains);
      }
    } catch (Throwable t) { /* intentionally unhandled */ }

    try {
      ArrayList<Map>  globalProps =  (ArrayList<Map>) configDict.get("globalProperties");
      if (globalProps != null){
        for (Map prop: globalProps){
          String key = (String) prop.get("key");
          String value = (String) prop.get("value");
          boolean overrideExisting = (boolean) prop.get("overrideExisting");
          singularConfig.withGlobalProperty(key, value, overrideExisting);
        }
      }
    } catch (Throwable t) { /* intentionally unhandled */ }

    singularLinkHandler = new SingularLinkHandler() {
      @Override
      public void onResolved(SingularLinkParams params) {
        // The deep link destination, as configured in the Manage Links page
        String deeplink = params.getDeeplink();

        // The passthrough parameters added to the link, if any.
        String passthrough = params.getPassthrough();

        // Whether the link configured as a deferred deep link.
        boolean isDeferred = params.isDeferred();

        // Resolved deep link query parameters as map.
        Map<String, String> urlParameters = params.getUrlParameters();

        // Add deep link handling code here
        final Map<String, Object> linkParams = new HashMap<>();
        linkParams.put("deeplink", deeplink);
        linkParams.put("passthrough", passthrough);
        linkParams.put("isDeferred", isDeferred);
        linkParams.put("urlParameters", urlParameters);

        uiThreadHandler.post(new Runnable() {
          @Override
          public void run() {
            if (channel != null) {
              channel.invokeMethod("singularLinksHandlerName",linkParams);
            }
          }
        });
      }
    };

    if (mIntent != null) {
      int intentHash = mIntent.hashCode();
      if (intentHash != currentIntentHash) {
        currentIntentHash = intentHash;
        singularConfig.withSingularLink(mIntent, singularLinkHandler,(long) shortLinkResolveTimeOut);
      }
    }
    singularConfig.withSingularDeviceAttribution(new SingularDeviceAttributionHandler() {
      @Override
      public void onDeviceAttributionInfoReceived(Map<String, Object> deviceAttributionData) {
        JSONObject deviceAttribution = new JSONObject(deviceAttributionData);

        uiThreadHandler.post(new Runnable() {
          @Override
          public void run() {
            if (channel != null) {
              channel.invokeMethod("deviceAttributionCallbackName",deviceAttribution.toString());
            }
          }
        });
      }
    });

    try {
      String customSdid = (String) configDict.get("customSdid");
      if (customSdid != null) {
        singularConfig.withSDID(customSdid, new SingularConfig.SdidAccessorHandler() {
          @Override
          public void didSetSdid(String result) {
            uiThreadHandler.post(new Runnable() {
              @Override
              public void run() {
                if (channel != null) {
                  channel.invokeMethod("didSetSdidCallbackName", result);
                }
              }
            });
          }

          @Override
          public void sdidReceived(String result) {
            uiThreadHandler.post(new Runnable() {
              @Override
              public void run() {
                if (channel != null) {
                  channel.invokeMethod("sdidReceivedCallbackName", result);
                }
              }
            });
          }
        });
      }
    } catch (Throwable throwable) { /* unhandled */}

    Singular.init(mContext, singularConfig);
  }

  private void start(final MethodCall call, final Result result) {
    configDict = (Map) call.arguments;
    initSDK();
  }

  private void setCustomUserId(final MethodCall call, final Result result) {
    String customUserId = call.argument("customUserId");
    Singular.setCustomUserId(customUserId);
  }

  private void unsetCustomUserId(final MethodCall call, final Result result) {
    Singular.unsetCustomUserId();
  }

  private void setDeviceCustomUserId(final MethodCall call, final Result result) {
    String customUserId = call.argument("customUserId");
    Singular.setDeviceCustomUserId(customUserId);
  }

  private void event(final MethodCall call, final Result result) {
    String eventName = call.argument("eventName");
    Singular.event(eventName);
  }

  private void eventWithArgs(final MethodCall call, final Result result) {
    String eventName = call.argument("eventName");
    Map<String, Object> extra = call.argument("args");
    Singular.event(eventName, new JSONObject(extra).toString());
  }

  private void customRevenue(final MethodCall call, final Result result) {
    String currency = call.argument("currency");
    String eventName = call.argument("eventName");

    double amount = call.argument("amount");

    Singular.customRevenue(eventName, currency, amount);
  }

  private void customRevenueWithArgs(final MethodCall call, final Result result) {
    String currency = call.argument("currency");
    String eventName = call.argument("eventName");
    double amount = call.argument("amount");
    Map args = call.argument("attributes");

    Singular.customRevenue(eventName, currency, amount, args);
  }

  private void trackingOptIn(final MethodCall call, final Result result) {
    Singular.trackingOptIn();
  }

  private void trackingUnder13(final MethodCall call, final Result result) {
    Singular.trackingUnder13();
  }

  private void stopAllTracking(final MethodCall call, final Result result) {
    Singular.stopAllTracking();
  }

  private void resumeAllTracking(final MethodCall call, final Result result) {
    Singular.resumeAllTracking();
  }

  private void isAllTrackingStopped(final MethodCall call, final Result result) {
    result.success(Singular.isAllTrackingStopped());
  }

  private void limitDataSharing(final MethodCall call, final Result result) {
    boolean shouldLimitDataSharing =  call.argument("limitDataSharing");
    Singular.limitDataSharing(shouldLimitDataSharing);
  }

  private void getLimitDataSharing(final MethodCall call, final Result result) {
    result.success(Singular.getLimitDataSharing());
  }

  private void clearGlobalProperties(final MethodCall call, final Result result) {
    Singular.clearGlobalProperties();
  }

  private void unsetGlobalProperty(final MethodCall call, final Result result) {
    String key = call.argument("key");
    Singular.unsetGlobalProperty(key);
  }

  private void setGlobalProperty(final MethodCall call, final Result result) {
    String key = call.argument("key");
    String value = call.argument("value");
    boolean overrideExisting = call.argument("overrideExisting");
    result.success(Singular.setGlobalProperty(key, value, overrideExisting));
  }

  private void getGlobalProperties(final MethodCall call, final Result result) {
    result.success(Singular.getGlobalProperties());
  }

  private void setFCMDeviceToken(final MethodCall call, final Result result) {
    String fcmToken = call.argument("fcmToken");
    Singular.setFCMDeviceToken(fcmToken);
  }

  private void registerDeviceTokenForUninstall(final MethodCall call, final Result result) {
    String fcmToken = call.argument("deviceToken");
    Singular.setFCMDeviceToken(fcmToken);
  }

  private void setWrapperNameAndVersion(final MethodCall call, final Result result) {
    String name = call.argument("name");
    String version = call.argument("version");

    Singular.setWrapperNameAndVersion(name, version);
  }

  private void createReferrerShortLink(final MethodCall call, final Result result) {
    String baseLink = call.argument("baseLink");
    String referrerName = call.argument("referrerName");
    String referrerId = call.argument("referrerId");
    Map args = call.argument("args");
    JSONObject params = new JSONObject(args);

    Singular.createReferrerShortLink(baseLink,
            referrerName,
            referrerId,
            params,
            new ShortLinkHandler() {
              @Override
              public void onSuccess(final String link) {
                uiThreadHandler.post(new Runnable() {
                  @Override
                  public void run() {
                    final Map<String, Object> linkParams = new HashMap<>();
                    linkParams.put("data", link);
                    linkParams.put("error", null);
                    if (channel != null) {
                      channel.invokeMethod("shortLinkCallbackName",linkParams);
                    }
                  }
                });
              }

              @Override
              public void onError(final String error) {
                uiThreadHandler.post(new Runnable() {
                  @Override
                  public void run() {
                    final Map<String, Object> linkParams = new HashMap<>();
                    linkParams.put("data", null);
                    linkParams.put("error", error);
                    if (channel != null){
                      channel.invokeMethod("shortLinkCallbackName",linkParams);
                    }
                  }
                });
              }
            });
  }
}

File: ios/Classes/SingularSDK.m
#import <Singular/Singular.h>
#import <Singular/SingularConfig.h>
#import <Singular/SingularLinkParams.h>
#import "SingularAppDelegate.h"
#import "SingularConstants.h"
#import "SingularSDK.h"

@implementation SingularSDK

static NSURL *tempOpenUrl;
static FlutterMethodChannel *channel;
static NSDictionary *configDict;

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
    channel = [FlutterMethodChannel methodChannelWithName:@"singular-api" binaryMessenger:[registrar messenger]];

    SingularSDK *instance = [[SingularSDK alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    if ([START isEqualToString:call.method]) {
        [self start:call withResult:result];
    } else if ([EVENT isEqualToString:call.method]) {
        [self event:call withResult:result];
    } else if ([EVENT_WITH_ARGS isEqualToString:call.method]) {
        [self eventWithArgs:call withResult:result];
    } else if ([SET_CUSTOM_USER_ID isEqualToString:call.method]) {
        [self setCustomUserId:call withResult:result];
    } else if ([UNSET_CUSTOM_USER_ID isEqualToString:call.method]) {
        [Singular unsetCustomUserId];
    } else if ([SET_DEVICE_CUSTOM_USER_ID isEqualToString:call.method]) {
        [self setDeviceCustomUserId:call withResult:result];
    } else if ([REGISTER_DEVICE_TOKEN_FOR_UNINSTALL isEqualToString:call.method]) {
        [self registerDeviceTokenForUninstall:call withResult:result];
    } else if ([CUSTOM_REVENUE isEqualToString:call.method]) {
        [self customRevenue:call withResult:result];
    } else if ([CUSTOM_REVENUE_WITH_ATTRIBUTES isEqualToString:call.method]) {
        [self customRevenueWithAttributes:call withResult:result];
    } else if ([CUSTOM_REVENUE_WITH_ALL_ATTRIBUTES isEqualToString:call.method]) {
        [self customRevenueWithAllAttributes:call withResult:result];
    } else if ([SET_WRAPPER_NAME_AND_VERSION isEqualToString:call.method]) {
        [self setWrapperNameAndVersion:call withResult:result];
    } else if ([GET_GLOBAL_PROPERTIES isEqualToString:call.method]) {
        result([Singular getGlobalProperties]);
    } else if ([SET_GLOBAL_PROPERTY isEqualToString:call.method]) {
        [self setGlobalProperty:call withResult:result];
    } else if ([UNSET_GLOBAL_PROPERTY isEqualToString:call.method]) {
        [self unsetGlobalProperty:call withResult:result];
    } else if ([CLEAR_GLOBAL_PROPERTIES isEqualToString:call.method]) {
        [Singular clearGlobalProperties];
    } else if ([TRACKING_OPT_IN isEqualToString:call.method]) {
        [Singular trackingOptIn];
    } else if ([TRACKING_UNDER13 isEqualToString:call.method]) {
        [Singular trackingUnder13];
    } else if ([STOP_ALL_TRACKING isEqualToString:call.method]) {
        [Singular stopAllTracking];
    } else if ([RESUME_ALL_TRACKING isEqualToString:call.method]) {
        [Singular resumeAllTracking];
    } else if ([IS_ALL_TRACKING_STOPPED isEqualToString:call.method]) {
        result(@([Singular isAllTrackingStopped]));
    } else if ([LIMIT_DATA_SHARING isEqualToString:call.method]) {
        [self limitDataSharing:call withResult:result];
    } else if ([GET_LIMIT_DATA_SHARING isEqualToString:call.method]) {
        result(@([Singular getLimitDataSharing]));
    } else if ([SKAN_REGISTER_APP_FOR_AD_ATTRIBUTION isEqualToString:call.method]) {
        [Singular skanRegisterAppForAdNetworkAttribution];
    } else if ([SKAN_UPDATE_CONVERSION_VALUE isEqualToString:call.method]) {
        [self skanUpdateConversionValue:call withResult:result];
    } else if ([SKAN_UPDATE_CONVERSION_VALUES isEqualToString:call.method]) {
        [self skanUpdateConversionValues:call withResult:result];
    } else if ([SKAN_GET_CONVERSION_VALUE isEqualToString:call.method]) {
        [self skanGetConversionValue:call withResult:result];
    } else if ([CREATE_REFERRER_SHORT_LINK isEqualToString:call.method]) {
        [self createReferrerShortLink:call withResult:result];
    } else {
        result(FlutterMethodNotImplemented);
    }
}

+ (void)initializeSingular {
    [SingularSDK initSDK];
}

+ (void)initSDK {
    if (configDict == nil) {
        return;
    }

    NSString *apiKey = configDict[@"apiKey"];
    NSString *secretKey = configDict[@"secretKey"];
    BOOL skAdNetworkEnabled = [configDict[@"skAdNetworkEnabled"] boolValue];
    BOOL clipboardAttribution = [configDict[@"clipboardAttribution"] boolValue];
    BOOL manualSkanConversionManagement = [configDict[@"manualSkanConversionManagement"] boolValue];
    int waitForTrackingAuthorizationWithTimeoutInterval = [configDict[@"waitForTrackingAuthorizationWithTimeoutInterval"] intValue];
    float shortLinkResolveTimeOut = [configDict[@"shortLinkResolveTimeOut"] floatValue];
    NSString *customUserId = configDict[@"customUserId"];

    SingularConfig *config = [[SingularConfig alloc] initWithApiKey:apiKey andSecret:secretKey];
    config.skAdNetworkEnabled = skAdNetworkEnabled;
    config.clipboardAttribution = clipboardAttribution;
    config.manualSkanConversionManagement = manualSkanConversionManagement;
    config.waitForTrackingAuthorizationWithTimeoutInterval = waitForTrackingAuthorizationWithTimeoutInterval;
    config.shortLinkResolveTimeOut = shortLinkResolveTimeOut;
    NSArray *props = configDict[@"globalProperties"];

    if (props != nil) {
        for (NSDictionary *prop in props) {
            NSString *key = [prop objectForKey:@"key"];
            NSString *value = [prop objectForKey:@"value"];
            BOOL overrideExisting = [[prop objectForKey:@"overrideExisting"]boolValue];
            [config setGlobalProperty:key withValue:value overrideExisting:overrideExisting];
        }
    }

    if (customUserId) {
        [Singular setCustomUserId:customUserId];
    }

    NSNumber *limitDataSharing = configDict[@"limitDataSharing"];

    if (![limitDataSharing isEqual:[NSNull null]]) {
        [Singular limitDataSharing:[limitDataSharing boolValue]];
    }

    NSNumber *sessionTimeout = configDict[@"sessionTimeout"];

    if (sessionTimeout >= 0) {
        [Singular setSessionTimeout:[sessionTimeout intValue]];
    }

    config.singularLinksHandler = ^(SingularLinkParams *params) {
        NSMutableDictionary *linkParams = [[NSMutableDictionary alloc] init];
        [linkParams setValue:[params getDeepLink] forKey:@"deeplink"];
        [linkParams setValue:[params getPassthrough] forKey:@"passthrough"];
        [linkParams setValue:@([params isDeferred]) forKey:@"isDeferred"];
        [linkParams setValue:([params getUrlParameters] ? [params getUrlParameters] : @{ }) forKey:@"urlParameters"];

        [channel invokeMethod:@"singularLinksHandlerName" arguments:linkParams];
    };

    if ([SingularAppDelegate shared].launchOptions != nil) {
        config.launchOptions = [SingularAppDelegate shared].launchOptions;
    } else if ([SingularAppDelegate shared].userActivity != nil) {
        config.userActivity = [SingularAppDelegate shared].userActivity;
    } else if ([SingularAppDelegate shared].openURL != nil) {
        config.openUrl = [SingularAppDelegate shared].openURL;
    } else {
        NSLog(@"everything is null");
    }

    config.conversionValueUpdatedCallback = ^(NSInteger conversionValue) {
        NSString *conversionValueUpdatedCallback = configDict[@"conversionValueUpdatedCallback"];

        if (conversionValueUpdatedCallback != nil) {
            [channel invokeMethod:@"conversionValueUpdatedCallbackName" arguments:@(conversionValue)];
        }
    };

    config.conversionValuesUpdatedCallback = ^(NSNumber *conversionValue, NSNumber *coarse, BOOL lock) {
        NSString *conversionValuesUpdatedCallback = configDict[@"conversionValuesUpdatedCallback"];

        if (conversionValuesUpdatedCallback != nil) {
            NSMutableDictionary *updatedConversionValues = [[NSMutableDictionary alloc] init];
            [updatedConversionValues setValue:(conversionValue != nil) ? @([conversionValue integerValue]) : @(-1) forKey:@"conversionValue"];
            [updatedConversionValues setValue:(coarse != nil) ? @([coarse integerValue]) : @(-1) forKey:@"coarse"];
            [updatedConversionValues setValue:@(lock) forKey:@"lock"];

            [channel invokeMethod:@"conversionValuesUpdatedCallbackName" arguments:updatedConversionValues];
        }
    };
    [Singular start:config];
}

- (void)start:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    configDict = call.arguments;
    [SingularSDK initSDK];
}

- (void)event:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString *eventName =  call.arguments[@"eventName"];

    [Singular event:eventName];
}

- (void)eventWithArgs:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString *eventName =  call.arguments[@"eventName"];
    NSDictionary *args = call.arguments[@"args"];

    [Singular event:eventName withArgs:args];
}

- (void)setCustomUserId:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString *customUserId = call.arguments[@"customUserId"];

    [Singular setCustomUserId:customUserId];
}

- (void)setDeviceCustomUserId:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString *customUserId = call.arguments[@"customUserId"];

    [Singular setDeviceCustomUserId:customUserId];
}

- (void)registerDeviceTokenForUninstall:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString *deviceToken = call.arguments[@"deviceToken"];

    [Singular registerDeviceTokenForUninstall:[deviceToken dataUsingEncoding:NSUTF8StringEncoding]];
}

- (void)customRevenue:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString *eventName =  call.arguments[@"eventName"];
    NSString *currency =  call.arguments[@"currency"];
    double amount = [call.arguments[@"amount"] doubleValue];

    [Singular customRevenue:eventName currency:currency amount:amount];
}

- (void)customRevenueWithAttributes:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString *eventName =  call.arguments[@"eventName"];
    NSString *currency =  call.arguments[@"currency"];
    double amount = [call.arguments[@"amount"] doubleValue];
    NSDictionary *attributes = call.arguments[@"attributes"];

    [Singular customRevenue:eventName currency:currency amount:amount withAttributes:attributes];
}

- (void)customRevenueWithAllAttributes:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString *eventName =  call.arguments[@"eventName"];
    NSString *currency =  call.arguments[@"currency"];
    double amount = [call.arguments[@"amount"] doubleValue];
    NSString *productSKU = call.arguments[@"productSKU"];
    NSString *productName = call.arguments[@"productName"];
    NSString *productCategory = call.arguments[@"productCategory"];
    int productQuantity = [call.arguments[@"productQuantity"] intValue];
    double productPrice = [call.arguments[@"productPrice"] doubleValue];

    [Singular customRevenue:eventName currency:currency amount:amount productSKU:productSKU productName:productName productCategory:productCategory productQuantity:productQuantity productPrice:productPrice];
}

- (void)createReferrerShortLink:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString *baseLink =  call.arguments[@"baseLink"];
    NSString *referrerName =  call.arguments[@"referrerName"];
    NSString *referrerId =  call.arguments[@"referrerId"];
    NSDictionary *args = call.arguments[@"args"];

    [Singular createReferrerShortLink:baseLink
                         referrerName:referrerName
                           referrerId:referrerId
                    passthroughParams:args
                    completionHandler:^(NSString *data, NSError *error) {
        NSMutableDictionary *linkParams = [[NSMutableDictionary alloc] init];

        if (data != nil) {
            [linkParams setValue:data
                          forKey:@"data"];
        }

        if (error != nil) {
            [linkParams setValue:error.description
                          forKey:@"error"];
        }

        [channel invokeMethod:@"shortLinkCallbackName"
                    arguments:linkParams];
    }];
}

- (void)setWrapperNameAndVersion:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString *wrapperName =  call.arguments[@"name"];
    NSString *version =  call.arguments[@"version"];

    [Singular setWrapperName:wrapperName andVersion:version];
}

- (void)setGlobalProperty:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString *key =  call.arguments[@"key"];
    NSString *value =  call.arguments[@"value"];
    BOOL overrideExisting =  [call.arguments[@"overrideExisting"] boolValue];

    result(@([Singular setGlobalProperty:key andValue:value overrideExisting:overrideExisting]));
}

- (void)unsetGlobalProperty:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString *key =  call.arguments[@"key"];

    [Singular unsetGlobalProperty:key];
}

- (void)limitDataSharing:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    BOOL shouldLimitDataSharing =  [call.arguments[@"limitDataSharing"] boolValue];

    [Singular limitDataSharing:shouldLimitDataSharing];
}

- (void)skanUpdateConversionValue:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString *conversionValue =  call.arguments[@"conversionValue"];

    if ([self isFieldValid:conversionValue]) {
        result(@([Singular skanUpdateConversionValue:[conversionValue integerValue]]));
    }
}

- (void)skanUpdateConversionValues:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSString *conversionValue =  call.arguments[@"conversionValue"];
    NSString *coarse =  call.arguments[@"coarse"];
    BOOL lock =  [call.arguments[@"lock"] boolValue];

    [Singular skanUpdateConversionValue:[conversionValue integerValue] coarse:[coarse integerValue] lock:lock];
}

- (void)skanGetConversionValue:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    result([Singular skanGetConversionValue]);
}

- (BOOL)isFieldValid:(NSObject *)field {
    if (field == nil) {
        return NO;
    }

    // Check if its an instance of the singleton NSNull.
    if ([field isKindOfClass:[NSNull class]]) {
        return NO;
    }

    // If field can be converted to a string, check if it has any content.
    NSString *str = [NSString stringWithFormat:@"%@", field];

    if (str != nil) {
        if ([str length] == 0) {
            return NO;
        }
    }

    return YES;
}

@end
